[AArch64][SVE] Enable max vector bandwidth for SVE

huntergr-arm opened the pull request:
Returns true for shouldMaximizeVectorBandwidth when the register type
is a scalable vector and SVE or streaming SVE are available.

huntergr-arm committed:
Enable max vector bandwidth for SVE

huntergr-arm committed:
Fix formatting

SamTebbs33 reviewed (approved):
Looks good and simple to me, thanks.

paulwalker-arm commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -333,8 +333,10 @@ AArch64TTIImpl::getInlineCallPenalty(const Function *F, const CallBase &Call,
   assert(K != TargetTransformInfo::RGK_Scalar);
-  return (K == TargetTransformInfo::RGK_FixedWidthVector &&
-          ST->isNeonAvailable());
+  return ((K == TargetTransformInfo::RGK_FixedWidthVector &&
+           ST->isNeonAvailable()) ||
+          (K == TargetTransformInfo::RGK_ScalableVector &&
+           ST->isSVEorStreamingSVEAvailable()));
```
At this stage can this be just `return true;`?

The original feature check was likely to block max bandwidth when using SVE for fixed length vectors. However, if SVE is good for max bandwidth scalable vectors, then I would expect it to be good enough for fixed length as well.

huntergr-arm replied:
Doing that causes the vectorizer to maximize bandwidth in streaming-compatible-sve-no-maximize-bandwidth.ll, which was explicitly testing against that. @sdesmalen-arm how important is this reduced VF?

I guess we could do something like `return !ST->isStreaming() && !ST->isStreamingCompatible()` ?

davemgreen commented:
My understanding is this is a pretty large change to make, changing the chosen vector factor for a lot of vectorized loops. We were quite careful when doing it for Neon to make sure the performance was OK overall, and it had decent theory behind it. We implemented a number of fixes and improvements to make sure that the performance for larger vector sizes was acceptable.

Do you have performance results for SVE? Is the main reason for dot vectorization? SVE has a different vectorization scheme in general where it relies more on top/bottom vectorization (which are not currently supported very much) and extending load / truncating stores.

huntergr-arm commented:
> My understanding is this is a pretty large change to make, changing the chosen vector factor for a lot of vectorized loops. We were quite careful when doing it for Neon to make sure the performance was OK overall, and it had decent theory behind it. We implemented a number of fixes and improvements to make sure that the performance for larger vector sizes was acceptable.
> 
> Do you have performance results for SVE? Is the main reason for dot vectorization? SVE has a different vectorization scheme in general where it relies more on top/bottom vectorization (which are not currently supported very much) and extending load / truncating stores.

I've run spec2017 on neoverse-v1 hardware so far, and I didn't see too much difference (besides parest failing to build, which I've fixed). Any suggestions for other benchmarks/platforms to check?

This is indeed initially targeted at enabling the dot product work, but the partial reduction intrinsics are intended to support other patterns as well in the future, including the top/bottom extending instructions.

Our other approach for dot products (if this proves to be too much of a change) would be to enable max vector bandwidth when an integer add reduction is present.

SamTebbs33 reviewed (approved):
With your latest testing on various benchmarks I'm happy to give this another LGTM.

paulwalker-arm reviewed (approved):
Immature SVE code generation was the primary reason to strict this when it was enabled for NEON.  I know you've tested this privately so functionally we should be good.  Based on this I'm happy to give it a try and see what falls out from wider community testing.

huntergr-arm committed:
Fix up test that landed after initial PR

Leporacanthicus commented:
In our Flang benchmarks, we're seeing ~6.5% worse performance in Spec-17 cam4_r, which we suspect is this change.

ytmukai commented:
We aim to enable this change by incorporating conservative criteria to mitigate the risk of performance degradation. Through several analyses, we identified two scenarios that could potentially lead to performance issues:

- Cases requiring mask packing/unpacking
- Cases where vector packing/unpacking causes bottlenecks in the computation pipeline

By enabling vector width maximization only when these conditions are not met, we believe we can avoid many performance regressions. We welcome any feedback or advice.

Details of the analysis are as follows:

Regarding mask packing/unpacking, the SPEC cam4_r case was found to be affected by this issue. Extracting the loop that experienced performance degradation, we have:

https://godbolt.org/z/jshbPoheT

```c
// Extracted from cam4_r vert_interp routine
void f(int n, float *restrict a, double *b, double *c, double *d) {
  for (int i = 0; i < n; i++) {
    if ((b[i] < c[i]) & (c[i] < d[i]))
      a[i] = n;
  }
}
```

When vectorized with `vscale x 4`, packing instructions are required to use a mask created with double (vscale x 2) in float (vscale x 4). However, the cost model does not account for these instructions, leading to an underestimation of the cost. (In type conversion cases, the cost of packing/unpacking instructions is represented as the cost of type conversion instructions.)

The actual execution results on Neoverse V2 with `n=512` are as follows: (Compiled with `-mcpu=fujitsu-monaka` to prevent ASIMD selected.)

|                                       | Selected VF | Cost Model         | #Cycles | #Instructions |
|---------------------------------------|-------------|--------------------|--------:|--------------:|
| Default                               |  vscale x 2 | Cost=9 (2.2/lane)  |     512 |          2463 |
| -mllvm -vectorizer-maximize-bandwidth |  vscale x 4 | Cost=14 (1.8/lane) |     589 |          2529 |

Although the cost model suggests vscale x 4 is superior, actual execution shows increased time and instruction count. In the case of `vscale x 4`, 30% of the instructions involve mask packing.

Excluding cases with potential mask packing/unpacking requirements, such as if statements, can prevent performance degradation due to incorrect VF selection.

Additionally, we found that performance can degrade even without an increase in instruction count. The following loop is an example:

https://godbolt.org/z/Gr6rEevbP

```c
void f(int n, short *restrict a, long *b, double *c) {
  for (int i = 0; i < n; i++) {
    a[i] = b[i] + c[i];
  }
}
```

The execution results on Neoverse V2 with `n=512` are as follows:

|                                       | Selected VF | Cost Model         | #Cycles | #Instructions | #Loads/Stores | #Vector Ops |
|---------------------------------------|-------------|--------------------|--------:|--------------:|--------------:|------------:|
| Default                               |  vscale x 2 | Cost=8 (2.0/lane)  |     286 |          2207 |           384 |         384 |
| -mllvm -vectorizer-maximize-bandwidth |  vscale x 8 | Cost=26 (1.6/lane) |     451 |          1951 |           288 |         480 |

In this case, packing/unpacking instructions are considered in the cost model, and the actual instruction count for `vscale x 8` is lower. However, the increased pressure on the vector computation pipeline due to packing/unpacking causes a bottleneck, resulting in performance degradation. To avoid this, adopting a larger VF only when vector computation costs improve could be considered. (Ideally, the cost model should represent pipeline types.)

We tested the performance by randomly selecting from the following types for each variable in the loop above: `char`, `short`, `int`, `long`, `__fp16`, `_Float16`, `float`, and `double`. With `-vectorizer-maximize-bandwidth` enabled, out of 100 cases, execution time increased by more than 10% in 18 cases and decreased by more than 10% in 14 cases. By limiting to cases where vector computation costs do not increase, the number of cases with a 10% increase dropped to 3, while those with a 10% decrease remained at 12.

ytmukai commented:
Hi @huntergr-arm,

Thank you for your explanation and work!

> We might be able to use some of the widening+operation instructions

In the former case, regarding the packing/unpacking of masks, it might be appropriate to account for costs in that manner. However, in the latter case, the problem lies in comparing costs using a single value, and addressing it in that manner will not resolve the issue. I have submitted a patch attempting to address this issue. (#156012)
